[NNC] Add fusion support of aten::to

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48975 [NNC] add fusion support for nan_to_num
* **#48976 [NNC] Add fusion support of aten::to**
* #48974 [NNC] add support for masked_fill
* #48973 [NNC] Add Support For is_nan

Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

Commit:
Add fusion support of aten::to

[ghstack-poisoned]

Commit:
Update on "Add fusion support of aten::to"




[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -87,6 +87,10 @@ bool isSupported(Node* node) {
       "aten::lerp.Tensor(Tensor self, Tensor end, Tensor weight) -> Tensor",
+      "aten::to.dtype(Tensor self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor",
+      "aten::to.device(Tensor self, Device device, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor",
+      "aten::to.dtype_layout(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, "
```
lol, I almost made this missing-comma mistake too yesterday.  Thanks clang-tidy :)

bertmaher replied:
oops nm this is intentional.

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

bertmaher reviewed (commented):
I'm a bit worried about the behavior when `copy=False` (the default).  Can fusing this op lead to some aliasing that we wouldn't handle appropriately?

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -849,8 +855,28 @@ void LLVMCodeGenImpl::visit(const Cast* v) {
     if (dstType->isFPOrFPVectorTy()) {
+      // as with eager, convert from Double -> Half by Converting to Float then
+      // Half
+      // TODO: __truncdfhf2
```
cc: @penguinwu, you saw this error the other day, this commit fixes it

penguinwu replied:
Though my PR triggered the undefined symbol. It also revealed another problem, where `half` was promoted to `double` instead of `float`. I am checking in a fix for the correct type promotion. 

But I have also been wondering why the promotion to double would cause an undefined symbol. It's good to know that that particular problem is fixed too!

eellison commented:
@bertmaher the aliasing will only occur if it's a No-op, which i've disabled as the output of a Fusion Group so I think we're fine

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -534,6 +538,17 @@ class TensorExprFuser {
+      return false;
+    }
+
+    return *n->input(0)->type()->expect<TensorType>() ==
```
Can we optimize away these ops with const-prop or something?

eellison replied:
sadly not, because we only have the profiled type here. If it is an a fusion group it will be a No-op tho.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -774,6 +775,15 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
 
+    case aten::to: {
+      auto node = v->node();
+      return computeOneOperand("aten_to", v, [node](const ExprHandle& a) {
```
AFAIU, `to` might have more than one input, and we seem to be ignoring them here. Should we specially handle or reject such cases?

eellison replied:
I think we're already rejecting them. I didn't add support for the aten::to overload which takes two Tensors.

eellison replied:
I could add support for it as a follow up, I think it's less common

ZolotukhinM replied:
I didn't mean `to` with two tensors, I mean `to` with optional arguments, like
```
x = y.to(torch.float32, non_blocking=True, copy=True, memory_format=torch.preserve_format)
```
My question was about these cases (which are missing in tests too).

eellison replied:
copy doesn't apply, we skip fusion in the case that the value is used, which there is a test for. 

memory_format doesn't apply, it's only applied to output tensors as a last step, just as with all other operations (at least when https://github.com/pytorch/pytorch/pull/48264 lands, which i'm trying to). 

and non_blocking i also don't think really applies, but i'm not really sure why it would ever be used or matter

ZolotukhinM replied:
What about `pin_memory` in the `to.dtype_layout`? Either way, please add a comment why it is safe to ignore the other arguments. Might be worth adding a test for these too.

eellison replied:
You're right about pin_memory, I think that's a cpu option to make it faster to copy to gpu. In this case, i'll skip. And agreed about adding comments/tests

ZolotukhinM reviewed (approved):
Looks good to me if you're certain that the semantics is properly preserved in all combos of arguments!

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -87,6 +87,10 @@ bool isSupported(Node* node) {
       "aten::lerp.Tensor(Tensor self, Tensor end, Tensor weight) -> Tensor",
+      "aten::to.dtype(Tensor self, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor",
+      "aten::to.device(Tensor self, Device device, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor",
+      "aten::to.dtype_layout(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None"
```
`// NOLINTNEXTLINE` to silence clang-tidy ;)

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]

Commit:
Update on "[NNC] Add fusion support of aten::to"


Add support for same device `torch.to` when the non-tensor inputs are constant

Differential Revision: [D25413164](https://our.internmc.facebook.com/intern/diff/D25413164)

[ghstack-poisoned]
